### PR TITLE
[c10d] Move DDP queue_reduction to C++

### PR DESCRIPTION
**Pull request by teng-li**
fully working version by using continuing on @goldsborough 's initial version.

waiting on the stream guard to be merged before adding more stream perf logics into the c++ version

### CODE REVIEW

**Review by pietern — COMMENTED** [review]
Can you change the test preamble to use the file store instead? In #12782 I changed all of them in test_c10d.py to use a file store to reduce the TCPStore bind address error induced test flakiness.

File: test/test_c10d.py [hunk 1]
```diff
@@ -828,6 +828,33 @@ def test_fp16(self):
             any(torch.isinf(p.grad).any() for p in ddp_model.parameters())
         )
 
+    def test_queue_reduction(self):
```
[the remark this answers is not in the document]
Comment (pietern) [reply test/test_c10d.py]: We can still build without NCCL if we build without CUDA no?

[the remark this answers is not in the document]
Comment (pietern) [reply test/test_c10d.py]: Nevermind, I see you added it.

**Review by pietern — APPROVED** [review]
Looking good! Any observable difference in perf?

File: test/test_c10d.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (teng-li) [reply test/test_c10d.py]: No, we won't build NCCL if CUDA is not enabled.

**Pull request closed without merging**